README.md
[![Build Status](https://travis-ci.org/johnjohndoe/Overpass.svg)](https://travis-ci.org/johnjohndoe/Overpass) [![Download](https://api.bintray.com/packages/tbsprs/maven/Overpass/images/download.svg)](https://bintray.com/tbsprs/maven/Overpass/_latestVersion) [![Apache License](http://img.shields.io/badge/license-Apache%20License%202.0-lightgrey.svg)](http://choosealicense.com/licenses/apache-2.0/) [![VersionEye](https://www.versioneye.com/user/projects/5658152672d5f3000900bf0c/badge.svg)](https://www.versioneye.com/user/projects/5658152672d5f3000900bf0c)

# Overpass library

A Java library containing a parser and models for the Overpass-Turbo API available here:

* http://overpass-turbo.eu


## Usage

The following example shows how to query post boxes:

```java
Map<String, String> tags = new HashMap<String, String>() {
    {
        put("amenity", "post_box");
    }
};
NodesQuery nodesQuery = new NodesQuery(600, 52.516667, 13.383333, tags, true, 13);
Call<OverpassResponse> streamsResponseCall = streamsService.getOverpassResponse(
	nodesQuery.getFormattedDataQuery());
// Execute streamsResponse call to send a request to the Overpass-Turbo API.
```

### Gradle build

To deploy the library to your local Maven repository run the following task:

```bash
$ ./gradlew publishToMavenLocal
```

Then, to use the library in your project add the following to
your top level `build.gradle`:

```
allprojects {
    repositories {
        jcenter()
        mavenLocal()
    }
}
```

and to your application module `build.gradle`:


```groovy
dependencies {
    compile 'info.metadude.java.library.overpass:overpass-library:${version}'
}
```

## Tests

Run the following command to execute all tests:

```groovy
$ ./gradlew clean test
```

## Which applications are using this library?

* WhereWhat Android Wear app, [sources][wherewhat-sources]


## Author

* [Tobias Preuss][tobias-preuss]

## License

    Copyright 2015 - 2016 Tobias Preuss

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


[tobias-preuss]: https://github.com/johnjohndoe
[wherewhat-sources]: https://github.com/ligi/WhereWhat


CHANGELOG.md
# Overpass changelog

## [v.2.0.0](https://github.com/johnjohndoe/Overpass/releases/tag/v.2.0.0)

* Published: 2016-03-22

### Changes

* Migrate to Retrofit v.2.0.0 and OkHttp v.3.2.0.
  Please note that the ApiModule now allows to pass in the OkHttpClient used by the application.
* Rename DataQuery to NodesQuery since it only supports querying nodes.


## [v.1.0.0](https://github.com/johnjohndoe/Overpass/releases/tag/v.1.0.0)

* Published: 2015-12-01

### Changes

* This is the initial release. Have fun!


src/test/java/info/metadude/java/library/overpass/test/OverpassServiceTest.java
package info.metadude.java.library.overpass.test;

import info.metadude.java.library.overpass.ApiModule;
import info.metadude.java.library.overpass.OverpassService;
import info.metadude.java.library.overpass.models.Element;
import info.metadude.java.library.overpass.models.OverpassResponse;
import info.metadude.java.library.overpass.utils.NodesQuery;
import org.junit.Before;
import org.junit.Test;
import retrofit2.Call;
import retrofit2.Response;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.fail;

public class OverpassServiceTest {

    protected OverpassService streamsService;

    @Before
    public void initStreamService() {
        streamsService = ApiModule.provideOverpassService();
    }

    @Test
    public void testThatRealServerIsReachable() {
        int responseLimit = 13;
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        NodesQuery nodesQuery = new NodesQuery(
                1600,
                52.516667,
                13.383333,
                tags,
                true,
                responseLimit);
        Call<OverpassResponse> streamsResponseCall = streamsService.getOverpassResponse(
                nodesQuery.getFormattedDataQuery());

        try {
            Response<OverpassResponse> response = streamsResponseCall.execute();
            if (response.isSuccessful()) {
                OverpassResponse overpassResponse = response.body();
                // Expect at least some data
                assertThat(overpassResponse)
                        .isNotNull();
                List<Element> elements = overpassResponse.elements;
                assertThat(elements)
                        .isNotNull()
                        .isNotEmpty();
                assertThat(elements.size()).isLessThanOrEqualTo(responseLimit);
                for (Element element : elements) {
                    testElement(element);
                }
            } else {
                fail("GetOffers response is not successful.");
            }
        } catch (IOException e) {
            fail("Should not throw " + e);
        }
    }

    private void testElement(Element element) {
        assertThat(element).isNotNull();
        assertThat(element.id).isGreaterThan(0);
        assertThat(element.lat).isGreaterThan(0);
        assertThat(element.lon).isGreaterThan(0);
        assertThat(element.type).isEqualTo("node");
        assertThat(element.tags).isNotNull();
    }

}


src/test/java/info/metadude/java/library/overpass/utils/test/DataQueryTest.java
package info.metadude.java.library.overpass.utils.test;

import info.metadude.java.library.overpass.utils.DataQuery;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

import static org.assertj.core.api.Assertions.assertThat;

public final class DataQueryTest {

    @Test
    public void testDataQueryWithTag() {
        String expectedDataQuery =
                "[out:json];node(around:600,52.516667,13.383333)" +
                        "[\"amenity\"=\"post_box\"];out qt 13;";
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        DataQuery dataQuery = new DataQuery(600, 52.516667, 13.383333, tags, true, 13);
        assertThat(dataQuery.getFormattedDataQuery()).isEqualTo(expectedDataQuery);
    }

    @Test
    public void testDataQueryWithInvalidRadius() {
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        int radius = 0;
        try {
            new DataQuery(radius, 52.516667, 13.383333, tags, true, 13);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected maximum radius to be greater then one, but was " + radius);
        }
    }

    @Test
    public void testDataQueryWithInvalidLongitude() {
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        int longitude = 0;
        try {
            new DataQuery(600, longitude, 13.383333, tags, true, 13);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected longitude between -180 and 180, but was " + longitude);
        }
    }

    @Test
    public void testDataQueryWithInvalidLatitude() {
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        int latitude = 0;
        try {
            new DataQuery(600, 52.516667, latitude, tags, true, 13);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected latitude between -90 and 90, but was " + latitude);
        }
    }

    @Test
    public void testDataQueryWithInvalidResponseCount() {
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        int maxResponseCount = 0;
        try {
            new DataQuery(600, 52.516667, 13.383333, tags, true, maxResponseCount);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected maximum response count to be " +
                    "greater then zero, but was " + maxResponseCount);
        }
    }

    @Test
    public void testUnsortedDataQuery() {
        String expectedDataQuery =
                "[out:json];node(around:600,52.516667,13.383333)" +
                        "[\"amenity\"=\"post_box\"];out 13;";
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        boolean sortByDistance = false;
        //noinspection ConstantConditions
        DataQuery dataQuery = new DataQuery(600, 52.516667, 13.383333, tags, sortByDistance, 13);
        assertThat(dataQuery.getFormattedDataQuery()).isEqualTo(expectedDataQuery);
    }

    @Test
    public void testDataQueryWithoutTags() {
        try {
            new DataQuery(600, 52.516667, 13.383333, null, true, 13);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected at least one tag.");
        }
    }

    @Test
    public void testDataQueryWithMultipleTags() {
        String expectedDataQuery =
                "[out:json];node(around:600,52.516667,13.383333)" +
                        "[\"amenity\"=\"recycling\"][\"recycling:clothes\"=\"yes\"];" +
                        "out qt 13;";
        Map<String, String> tags = new TreeMap<String, String>() {
            {
                put("amenity", "recycling");
                put("recycling:clothes", "yes");
            }
        };
        DataQuery dataQuery = new DataQuery(600, 52.516667, 13.383333, tags, true, 13);
        assertThat(dataQuery.getFormattedDataQuery()).isEqualTo(expectedDataQuery);
    }

}


src/test/java/info/metadude/java/library/overpass/utils/test/NodesQueryTest.java
package info.metadude.java.library.overpass.utils.test;

import info.metadude.java.library.overpass.utils.NodesQuery;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

import static org.assertj.core.api.Assertions.assertThat;

public final class NodesQueryTest {

    @Test
    public void testDataQueryWithTag() {
        String expectedDataQuery =
                "[out:json];node(around:600,52.516667,13.383333)" +
                        "[\"amenity\"=\"post_box\"];out qt 13;";
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        NodesQuery nodesQuery = new NodesQuery(600, 52.516667, 13.383333, tags, true, 13);
        assertThat(nodesQuery.getFormattedDataQuery()).isEqualTo(expectedDataQuery);
    }

    @Test
    public void testDataQueryWithInvalidRadius() {
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        int radius = 0;
        try {
            new NodesQuery(radius, 52.516667, 13.383333, tags, true, 13);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected maximum radius to be greater then one, but was " + radius);
        }
    }

    @Test
    public void testDataQueryWithInvalidLongitude() {
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        int longitude = 0;
        try {
            new NodesQuery(600, longitude, 13.383333, tags, true, 13);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected longitude between -180 and 180, but was " + longitude);
        }
    }

    @Test
    public void testDataQueryWithInvalidLatitude() {
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        int latitude = 0;
        try {
            new NodesQuery(600, 52.516667, latitude, tags, true, 13);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected latitude between -90 and 90, but was " + latitude);
        }
    }

    @Test
    public void testDataQueryWithInvalidResponseCount() {
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        int maxResponseCount = 0;
        try {
            new NodesQuery(600, 52.516667, 13.383333, tags, true, maxResponseCount);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected maximum response count to be " +
                    "greater then zero, but was " + maxResponseCount);
        }
    }

    @Test
    public void testUnsortedDataQuery() {
        String expectedDataQuery =
                "[out:json];node(around:600,52.516667,13.383333)" +
                        "[\"amenity\"=\"post_box\"];out 13;";
        Map<String, String> tags = new HashMap<String, String>() {
            {
                put("amenity", "post_box");
            }
        };
        boolean sortByDistance = false;
        //noinspection ConstantConditions
        NodesQuery nodesQuery = new NodesQuery(600, 52.516667, 13.383333, tags, sortByDistance, 13);
        assertThat(nodesQuery.getFormattedDataQuery()).isEqualTo(expectedDataQuery);
    }

    @Test
    public void testDataQueryWithoutTags() {
        try {
            new NodesQuery(600, 52.516667, 13.383333, null, true, 13);
        } catch (Exception e) {
            assertThat(e.getMessage()).isEqualTo("Expected at least one tag.");
        }
    }

    @Test
    public void testDataQueryWithMultipleTags() {
        String expectedDataQuery =
                "[out:json];node(around:600,52.516667,13.383333)" +
                        "[\"amenity\"=\"recycling\"][\"recycling:clothes\"=\"yes\"];" +
                        "out qt 13;";
        Map<String, String> tags = new TreeMap<String, String>() {
            {
                put("amenity", "recycling");
                put("recycling:clothes", "yes");
            }
        };
        NodesQuery nodesQuery = new NodesQuery(600, 52.516667, 13.383333, tags, true, 13);
        assertThat(nodesQuery.getFormattedDataQuery()).isEqualTo(expectedDataQuery);
    }

}


src/main/java/info/metadude/java/library/overpass/ApiModule.java
package info.metadude.java.library.overpass;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

public final class ApiModule {

    public static OverpassService provideOverpassService() {
        return provideOverpassService(null);
    }

    public static OverpassService provideOverpassService(final OkHttpClient okHttpClient) {
        return createRetrofit(OverpassService.BASE_URL, okHttpClient)
                .create(OverpassService.class);
    }

    private static Retrofit createRetrofit(final String baseUrl, final OkHttpClient okHttpClient) {
        OkHttpClient httpClient = okHttpClient;
        if (httpClient == null) {
            httpClient = new OkHttpClient();
        }

        return new Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(MoshiConverterFactory.create())
                .client(httpClient)
                .build();
    }

}



src/main/java/info/metadude/java/library/overpass/models/Element.java
package info.metadude.java.library.overpass.models;

import java.util.Map;

public class Element {

    public String type;

    public long id;

    public double lat;

    public double lon;

    public Map<String, String> tags;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Element element = (Element) o;
        return id == element.id && Double.compare(element.lat, lat) == 0 &&
                Double.compare(element.lon, lon) == 0 &&
                !(type != null ? !type.equals(element.type) : element.type != null) &&
                !(tags != null ? !tags.equals(element.tags) : element.tags != null);
    }

    @Override
    public int hashCode() {
        int result;
        long temp;
        result = type != null ? type.hashCode() : 0;
        result = 31 * result + (int) (id ^ (id >>> 32));
        temp = Double.doubleToLongBits(lat);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(lon);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        result = 31 * result + (tags != null ? tags.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "Element{" +
                "type='" + type + '\'' +
                ", id=" + id +
                ", lat=" + lat +
                ", lon=" + lon +
                ", tags=" + tags +
                '}';
    }

}


src/main/java/info/metadude/java/library/overpass/OverpassService.java
package info.metadude.java.library.overpass;

import info.metadude.java.library.overpass.models.OverpassResponse;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Query;

/**
 * Service to query Overpass-Turbo.
 * Example query: http://overpass-api.de/api/interpreter?data=[out:json];node(around:1600,52.516667,13.383333)["amenity"="post_box"];out qt 13;
 */
public interface OverpassService {

    String BASE_URL = "http://overpass-api.de";

    /**
     * Returns a Overpass response for the given query.
     *
     * @param data Overpass QL string data part
     *             Example: [out:json];node(around:1600,52.516667,13.383333)["amenity"="post_box"];out qt 13;
     * @return a call to execute the actual query.
     */
    @GET("/api/interpreter")
    Call<OverpassResponse> getOverpassResponse(
            @Query("data") String data
    );

}


src/main/java/info/metadude/java/library/overpass/utils/DataQuery.java
package info.metadude.java.library.overpass.utils;

import java.util.Map;

/**
 * Constructs a data query for Overpass-Turbo
 * Example: "[out:json];node(around:600,52.516667,13.383333)[\"amenity\"=\"post_box\"];out qt 13;"
 */
public class DataQuery {

    protected final int radius;

    protected final double latitude; // 52.516667

    protected final double longitude; // 13.383333

    protected final Map<String, String> tags;

    protected final boolean sortByDistance;

    protected final int maxResponseCount;

    public DataQuery(int radius,
                     double latitude,
                     double longitude,
                     Map<String, String> tags,
                     boolean sortByDistance,
                     int maxResponseCount) {
        this.radius = radius;
        this.latitude = latitude;
        this.longitude = longitude;
        this.tags = tags;
        this.sortByDistance = sortByDistance;
        this.maxResponseCount = maxResponseCount;
        validateParams();
    }

    public String getFormattedDataQuery() {
        String sortInstruction = sortByDistance ? "qt " : "";
        String formattedTags = getFormattedTags();
        return "[out:json];" +
                "node(around:" + radius + "," + latitude + "," + longitude + ")" +
                formattedTags +
                "out " +
                sortInstruction +
                maxResponseCount + ";";
    }

    private void validateParams() {
        if (radius < 2) {
            throw new IllegalArgumentException("Expected maximum radius to be greater then one, but was " + radius);
        }
        if (latitude < -90 || latitude > 90) {
            throw new IllegalArgumentException("Expected latitude between -90 and 90, but was " + latitude);
        }
        if (longitude < -180 || longitude > 180) {
            throw new IllegalArgumentException("Expected longitude between -180 and 180, but was " + longitude);
        }
        if (tags == null || tags.isEmpty()) {
            throw new IllegalArgumentException("Expected at least one tag.");
        }
        if (maxResponseCount < 1) {
            throw new IllegalArgumentException("Expected maximum response count to be " +
                    "greater then zero, but was " + maxResponseCount);
        }
    }

    private String getFormattedTags() {
        StringBuilder stringBuilder = new StringBuilder();
        for (String key : tags.keySet()) {
            stringBuilder
                    .append("[\"")
                    .append(key).append("\"=\"")
                    .append(tags.get(key))
                    .append("\"]");
        }
        stringBuilder.append(";");
        return stringBuilder.toString();
    }

}


src/main/java/info/metadude/java/library/overpass/utils/NodesQuery.java
package info.metadude.java.library.overpass.utils;

import java.util.Map;

/**
 * Constructs a data query limited to nodes for Overpass-Turbo
 * Example: "[out:json];node(around:600,52.516667,13.383333)[\"amenity\"=\"post_box\"];out qt 13;"
 */
public class NodesQuery {

    protected final int radius;

    protected final double latitude; // 52.516667

    protected final double longitude; // 13.383333

    protected final Map<String, String> tags;

    protected final boolean sortByDistance;

    protected final int maxResponseCount;

    public NodesQuery(int radius,
                      double latitude,
                      double longitude,
                      Map<String, String> tags,
                      boolean sortByDistance,
                      int maxResponseCount) {
        this.radius = radius;
        this.latitude = latitude;
        this.longitude = longitude;
        this.tags = tags;
        this.sortByDistance = sortByDistance;
        this.maxResponseCount = maxResponseCount;
        validateParams();
    }

    public String getFormattedDataQuery() {
        String sortInstruction = sortByDistance ? "qt " : "";
        String formattedTags = getFormattedTags();
        return "[out:json];" +
                "node(around:" + radius + "," + latitude + "," + longitude + ")" +
                formattedTags +
                "out " +
                sortInstruction +
                maxResponseCount + ";";
    }

    private void validateParams() {
        if (radius < 2) {
            throw new IllegalArgumentException("Expected maximum radius to be greater then one, but was " + radius);
        }
        if (latitude < -90 || latitude > 90) {
            throw new IllegalArgumentException("Expected latitude between -90 and 90, but was " + latitude);
        }
        if (longitude < -180 || longitude > 180) {
            throw new IllegalArgumentException("Expected longitude between -180 and 180, but was " + longitude);
        }
        if (tags == null || tags.isEmpty()) {
            throw new IllegalArgumentException("Expected at least one tag.");
        }
        if (maxResponseCount < 1) {
            throw new IllegalArgumentException("Expected maximum response count to be " +
                    "greater then zero, but was " + maxResponseCount);
        }
    }

    private String getFormattedTags() {
        StringBuilder stringBuilder = new StringBuilder();
        for (String key : tags.keySet()) {
            stringBuilder
                    .append("[\"")
                    .append(key).append("\"=\"")
                    .append(tags.get(key))
                    .append("\"]");
        }
        stringBuilder.append(";");
        return stringBuilder.toString();
    }

}
